### PR TITLE
[PREVIEW] PAY-824 : Add keyword to a fixed fee

### PR DESCRIPTION
**Pull request by tarun-palisetty**

### CODE REVIEW

File: model/src/main/java/uk/gov/hmcts/fees2/register/data/service/validator/FeeValidator.java [hunk 1]
```diff
@@ -71,4 +85,56 @@ private void setDefaultValues(Fee fee) {
...
+
+         if (fees.size() > 0) {
+             LOG.info("No. of fees found for the matching reference data: {}", fees.size());
+
+
+             switch (fee.getFeeType()) {
+                 case "FixedFee":
+                     return fees.stream().filter(f -> f.getFeeType().equals("FixedFee")).findAny().isPresent();
+                 default:
+                     return false;
```
Comment (smathangi) [inline model/src/main/java/uk/gov/hmcts/fees2/register/data/service/validator/FeeValidator.java]: then call feeRepository.findAll(..) for FixedFee type only?

File: model/src/main/java/uk/gov/hmcts/fees2/register/data/service/impl/FeeServiceImpl.java [hunk 2]
```diff
@@ -81,6 +84,10 @@
     public Fee save(Fee fee) {
         feeValidator.validateAndDefaultNewFee(fee);
 
+        if (feeValidator.isFeeExists(fee)) {
```
Comment (smathangi) [inline model/src/main/java/uk/gov/hmcts/fees2/register/data/service/impl/FeeServiceImpl.java]: it'll give a problem for concurrent requests. May be method synchronized / locks etc?

Comment (tarun-palisetty) [reply model/src/main/java/uk/gov/hmcts/fees2/register/data/service/impl/FeeServiceImpl.java]: I suppose even if we make method as synchronized it is still not going be concurrent proof, as the API runs in prod runs on multiple nodes.

If we still want to make it synchronized, ideally we should compromise with the performance.

File: api/src/main/java/uk/gov/hmcts/fees2/register/util/FeeFactory.java [hunk 3]
```diff
@@ -1,14 +1,13 @@
...
-import uk.gov.hmcts.fees2.register.api.contract.loader.request.LoaderFixedFeeDto;
 import uk.gov.hmcts.fees2.register.api.contract.request.*;
 import uk.gov.hmcts.fees2.register.data.model.*;
 
 public class FeeFactory {
 
     public static Fee getFee(FeeDto request){
         Fee fee = null;
-        if (request instanceof FixedFeeDto || request instanceof LoaderFixedFeeDto)
+        if (request instanceof FixedFeeDto)
```
Comment (jaleen) [inline api/src/main/java/uk/gov/hmcts/fees2/register/util/FeeFactory.java]: Please add a TODO here to fix Factory method. It shouldn't be using the instanceof

File: api/src/main/resources/db/changelog/db.changelog-fees-managed-v5.yaml [hunk 4]
```diff
@@ -163,3 +163,15 @@ databaseChangeLog:
...
             VALUES ('other',current_timestamp,current_timestamp) ON CONFLICT DO NOTHING
+
+- changeSet:
+      id: 1534341462258
+      author: tarun
+      changes:
+      - addColumn:
+            columns:
+            - column:
+                  name: keyword
```
Comment (jaleen) [inline api/src/main/resources/db/changelog/db.changelog-fees-managed-v5.yaml]: Add an index on keyword and add another index to include all six refrence data in same index if possible

File: api/src/test/java/uk/gov/hmcts/fees2/register/api/controllers/base/FeeDataUtils.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+import java.util.List;
+
+import static uk.gov.hmcts.fees2.register.api.contract.FeeVersionDto.feeVersionDtoWith;
+import static uk.gov.hmcts.fees2.register.api.contract.request.FixedFeeDto.fixedFeeDtoWith;
+import static uk.gov.hmcts.fees2.register.api.contract.request.RangedFeeDto.rangedFeeDtoWith;
+
+public class FeeDataUtils {
+
+    public static FixedFeeDto getCreateProbateCopiesFeeRequest() {
+        return fixedFeeDtoWith()
```
Comment (jaleen) [inline api/src/test/java/uk/gov/hmcts/fees2/register/api/controllers/base/FeeDataUtils.java]: good one

**Pull request merged**